### PR TITLE
EASY-2670: change state from a command line

### PR DESCRIPTION
**Pull request by jo-pol**
Fixes EASY-2670 change state from a command line

#### When applied it will
* 
* 
* 

#### How should this be manually tested?
* [x] image

#### Related pull requests on github
repo                       | PR                | note
-------------------------- | ----------------- | ----
easy-                      | [PR#](PRlink)     | ..

### CODE REVIEW

File: src/test/scala/nl.knaw.dans.easy.deposit/EasyDepositApiAppSpec.scala [hunk 1]
```diff
@@ -15,27 +15,93 @@
...
+  }
+
+  it should "alter the state" in {
+    val uuid = app.createDeposit(defaultUserInfo.id).getOrRecover(e => fail(e)).id // preparation
+
+    app.forceChangeState(defaultUserInfo.id, uuid, newState, update = true) shouldBe expectedStates
+    app.getDepositState(defaultUserInfo.id, uuid) shouldBe Success(newState) // post condition
+  }
+
+  it should "cause trouble when setting the state from draft to submitted" in {
```
Comment (jo-pol) [inline src/test/scala/nl.knaw.dans.easy.deposit/EasyDepositApiAppSpec.scala]: It is even more dangerous when the deposit has been submitted before. Then there will be `bag-store.bag-id` but that is related to the previous attempt.

File: src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala [hunk 2]
```diff
@@ -40,12 +45,32 @@ class CommandLineOptions(args: Array[String], configuration: Configuration) exte
...
+  val runService: Subcommand = new Subcommand("run-service") {
     descr(
       "Starts EASY Deposit Api as a daemon that services HTTP requests")
     footer(SUBCOMMAND_SEPARATOR)
   }
+  val changeState = new Subcommand("change-state") {
+    descr(
+      "Changes the state of a deposit, when changing to SUBMITTED just the state is changed, the rest of the submit-cycle is not started")
+    val doUpdate: ScallopOption[Boolean] = opt(name = "doUpdate", noshort = true, required = false,
+      descr = s"without this argument only the current status is shown in json format")
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala]: This is not what the omission of `--doUpdate` should do: it should run the same code as when `--doUpdate` is provided, but replace side effects (in this case 'saving the new state') with a printed message stating the side effect that would have taken place if `--doUpdate` would have been provided.

File: src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala [hunk 3]
```diff
@@ -40,12 +45,32 @@ class CommandLineOptions(args: Array[String], configuration: Configuration) exte
...
+    val state: ScallopOption[State] = opt(name = "label", short = 'l', required = true,
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala]:
```suggestion
    val stateLabel: ScallopOption[State] = opt(name = "label", short = 'l', required = true,
```

File: src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala [hunk 4]
```diff
@@ -40,12 +45,32 @@ class CommandLineOptions(args: Array[String], configuration: Configuration) exte
...
+      descr = s"The label of the new state, one of: ${State.values.mkString(", ")}")
+    val description: ScallopOption[String] = opt(name = "description", short = 'd', required = true,
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala]:
```suggestion
    val stateDescription: ScallopOption[String] = opt(name = "description", short = 'd', required = true,
```

File: src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala [hunk 5]
```diff
@@ -40,12 +45,32 @@ class CommandLineOptions(args: Array[String], configuration: Configuration) exte
...
+      descr = "A desription of the new state")
+    val draftOwnerId: ScallopOption[String] = trailArg(name = "draftOwnerId", required = true,
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala]:
```suggestion
    val draftOwnerId: ScallopOption[String] = trailArg(name = "ownerId", required = true,
```

File: src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala [hunk 6]
```diff
@@ -40,12 +45,32 @@ class CommandLineOptions(args: Array[String], configuration: Configuration) exte
...
+      descr = "The owner of the existing draft deposit")
+    val draftDepositId: ScallopOption[UUID] = trailArg(name = "draftDepositId", required = true,
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala]:
```suggestion
    val draftDepositId: ScallopOption[UUID] = trailArg(name = "depositId", required = true,
```

File: src/main/scala/nl.knaw.dans.easy.deposit/Command.scala [hunk 7]
```diff
@@ -27,21 +28,28 @@ object Command extends App with DebugEnhancedLogging {
...
   Try { new EasyDepositApiApp(configuration) }
     .flatMap(runSubcommand)
     .doIfSuccess(msg => Console.err.println(s"OK: $msg"))
     .doIfFailure { case e => logger.error(e.getMessage, e) }
     .doIfFailure { case NonFatal(e) => Console.err.println(s"FAILED: ${ e.getMessage }") }
 
-  private def runSubcommand(app: EasyDepositApiApp): Try[FeedBackMessage] = {
-    commandLine.subcommand
+  def runSubcommand(clo: CommandLineOptions, app: EasyDepositApiApp): Try[FeedBackMessage] =
+    clo.subcommand
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/Command.scala]: why is this change necessary?

File: src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala (hunk 5 above)
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala]: `trailArg` has `required = true` as default. no need to state it explicitly.
same below

File: src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala [hunk 8]
```diff
@@ -210,12 +211,29 @@ class EasyDepositApiApp(configuration: Configuration) extends DebugEnhancedLoggi
...
+    getDepositStateManager(user, id)
+      .flatMap(_.getStateInfo)
+  }
+
+  def forceChangeState(user: String, id: UUID, newStateInfo: StateInfo, update: Boolean): Try[String] = {
+    trace(user, id)
+    for {
+      stateManager <- getDepositStateManager(user, id)
+      oldStateInfo <- stateManager.getStateInfo
+      _ = if (update) stateManager.saveInDraft(newStateInfo)
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala]: `else` ... ???
See earlier comment

File: src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala [hunk 9]
```diff
@@ -210,12 +211,29 @@ class EasyDepositApiApp(configuration: Configuration) extends DebugEnhancedLoggi
...
+    } yield
+      s"""OLD: ${ toJson(oldStateInfo) }
+         |NEW: ${ toJson(newStateInfo) }
+         |""".stripMargin
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala]: Not sure if JSON is the best format to be used here.

File: src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala (hunk 8 above)
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala]: `else` ... ???
See earlier comment

File: src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala [hunk 10]
```diff
@@ -210,12 +211,29 @@ class EasyDepositApiApp(configuration: Configuration) extends DebugEnhancedLoggi
...
+    trace(user, id)
+    getDepositStateManager(user, id)
+      .flatMap(_.getStateInfo)
+  }
+
+  def forceChangeState(user: String, id: UUID, newStateInfo: StateInfo, update: Boolean): Try[String] = {
+    trace(user, id)
+    for {
+      stateManager <- getDepositStateManager(user, id)
+      oldStateInfo <- stateManager.getStateInfo
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala]: isn't this the same as `getDepositState` above?

File: pom.xml [hunk 11]
```diff
@@ -28,7 +28,7 @@
     <version>1.2.1-SNAPSHOT</version>
     <name>EASY Deposit API</name>
     <url>https://github.com/DANS-KNAW/easy-deposit-api</url>
-    <description>JSON-based deposit service.</description>
+    <description>This service offers a JSON-based deposit API</description>
```
Comment (rvanheest) [inline pom.xml]: shouldn't this be the same as the `description` in `CommandLineOptions`? there should be a test for this, right? Or at least it is the case in other modules; not sure about this 😉

Comment (jo-pol) [reply pom.xml]: The test whether the description of the `CommandLineOptions` is found in the pom. The pom can have extra's.

File: src/main/scala/nl.knaw.dans.easy.deposit/Command.scala (hunk 7 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.deposit/Command.scala]: whoops, did not revert enough.

File: src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala [hunk 12]
```diff
@@ -40,12 +45,32 @@ class CommandLineOptions(args: Array[String], configuration: Configuration) exte
...
+
+  val runService: Subcommand = new Subcommand("run-service") {
     descr(
       "Starts EASY Deposit Api as a daemon that services HTTP requests")
     footer(SUBCOMMAND_SEPARATOR)
   }
+  val changeState = new Subcommand("change-state") {
+    descr(
+      "Changes the state of a deposit, when changing to SUBMITTED just the state is changed, the rest of the submit-cycle is not started")
+    val doUpdate: ScallopOption[Boolean] = opt(name = "doUpdate", noshort = true, required = false,
```
[the remark this answers is not in the document]
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala]: The tests show that the default is false.

File: src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala (hunk 5 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.deposit/CommandLineOptions.scala]: now you are inconsistent with wanting an explicit default for the `opt[Boolean]`

File: src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala (hunk 10 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala]: indeed, but now I don't have to create the `DepositDir` and `StateManager` again for the update.

File: src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala (hunk 9 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala]: Then please specify what you want. See the unit tests.

File: docs/index.md [hunk 13]
```diff
@@ -22,14 +22,25 @@ For details about the service API see the [OpenAPI specification].
...
+      -v, --version   Show version of this program
+    
+    Subcommand: run-service - Starts EASY Deposit Api as a daemon that services HTTP requests
+      -h, --help   Show help message
+    ---
+    
+    Subcommand: change-state - Changes the state of a deposit, when changing to SUBMITTED just the state is changed, the rest of the submit-cycle is not started
+      -d, --description  <arg>   A desription of the new state
+          --doUpdate             without this argument only the current status is
+                                 shown in json format
```
Comment (rvanheest) [inline docs/index.md]: > without this argument only the current status is shown in json format

Is that really true? Hopefully it will also show the state you would have changed to if this argument was provided.

File: src/main/scala/nl.knaw.dans.easy.deposit/Command.scala [hunk 14]
```diff
@@ -27,22 +28,25 @@ object Command extends App with DebugEnhancedLogging {
...
-  }
+  val commandLine: CommandLineOptions = new CommandLineOptions(args, configuration)
   Try { new EasyDepositApiApp(configuration) }
     .flatMap(runSubcommand)
     .doIfSuccess(msg => Console.err.println(s"OK: $msg"))
     .doIfFailure { case e => logger.error(e.getMessage, e) }
     .doIfFailure { case NonFatal(e) => Console.err.println(s"FAILED: ${ e.getMessage }") }
 
-  private def runSubcommand(app: EasyDepositApiApp): Try[FeedBackMessage] = {
+  def runSubcommand(app: EasyDepositApiApp) =
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/Command.scala]: Please keep the change as minimal as possible!

```suggestion
  private def runSubcommand(app: EasyDepositApiApp): Try[FeedBackMessage] = {
```
